dep0/lib/03_typecheck/src/check.cpp

```diff
@@ -471,55 +288,61 @@ expected<expr_t> check_expr(context_t const& ctx, parser::expr_t const& x, sort_
                                 std::move(*rhs)}});
                 });
         },
-        [&] (parser::expr_t::var_t const& x) -> expected<expr_t>
+        [&] (parser::expr_t::init_list_t const& init_list) -> expected<expr_t>
         {
-            auto const var = expr_t::var_t{x.name};
-            auto const val = ctx[var];
-            if (not val)
-            {
-                std::ostringstream err;
-                pretty_print<properties_t>(err << "unknown variable `", var) << '`';
-                return error_t::from_error(dep0::error_t(err.str(), loc), ctx, expected_type);
-            }
-            auto const result = 
-                make_legal_expr(
-                    match(
-                        val->value,
-                        [&] (type_def_t const&) -> sort_t { return derivation_rules::make_typename(); },
-                        [&] (expr_t const& expr) -> sort_t { return expr.properties.sort.get(); }),
-                    var);
-            if (auto eq = is_beta_delta_equivalent(ctx, result.properties.sort.get(), expected_type))
-                return result;
-            else
-                return type_error(result.properties.sort.get(), std::move(eq.error()));
+            // initializer lists cannot be type-assigned without an expected type (which for now can only be an array)
```

Comment:
Unless we change the grammar to `ty=expr '{' values=... '}'` to make the type mandatory?

Reply:
But consider this (in fact I can add a test in this PR for this)

```
auto signed_or_unsigned(bool signed) -> typename
{
    if (signed) return i32_t; else return u32_t;
}

auto first(bool signed, array_t(signed_or_unsigned(signed), 3) xs) -> signed_or_unsigned(signed)
{
    return xs[0];
}

i32_t g()
{
    return first(true, {-1,-2,-3}); // here we would instead need to write `array_t(i32_t, 3){-1, -2, -3}`
}
```